### PR TITLE
Add support for RADIO models

### PR DESCRIPTION
**Pull request by meatybobby**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46425)

# What does this PR do?

This PR adds support for NVIDIA RADIO models

- [ ] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/radio/modular_radio.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+    def forward(self, pixel_values: torch.Tensor) -> torch.Tensor:
+        patches = self.patch_projection(self._image_to_patches(pixel_values))
+        input_dims = (pixel_values.shape[-2] // self.patch_size, pixel_values.shape[-1] // self.patch_size)
+        patches = patches + self._interpolate_position_embedding(input_dims, patches.dtype)
+        prefix = self.cls_register_token.unsqueeze(0).expand(patches.shape[0], -1, -1)
+        return torch.cat([prefix, patches], dim=1)
+
+
+class RADIOMLP(Dinov2MLP):
```
Comment (ArthurZucker) [inline src/transformers/models/radio/modular_radio.py]: In general if we can use `Radio` would be better!

**Review by ArthurZucker — APPROVED** [review]
LGTM I would like `RadioMLP` (lower cap)

File: src/transformers/models/radio/modular_radio.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            init.ones_(module.weight)
+        elif isinstance(module, RadioPatchEmbeddings):
+            init.trunc_normal_(module.position_embedding, mean=0.0, std=std)
+            init.trunc_normal_(module.cls_register_token, mean=0.0, std=std)
+        elif isinstance(module, RadioLayerScale):
+            init.constant_(module.lambda1, self.config.layerscale_value)
+
+
+@auto_docstring
+class RADIOModel(RadioPreTrainedModel):
```
Comment (ArthurZucker) [inline src/transformers/models/radio/modular_radio.py]:
```suggestion
class RadioModel(RadioPreTrainedModel):
```

File: src/transformers/models/radio/modular_radio.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    def patch_size(self) -> int:
+        return self.config.patch_size
+
+    def make_preprocessor_external(self):
+        """Detach the input conditioner (caller applies normalization itself)."""
+        conditioner = self.input_conditioner
+        self.input_conditioner = nn.Identity()
+        return conditioner
+
+    @auto_docstring
```
Comment (ArthurZucker) [inline src/transformers/models/radio/modular_radio.py:228]:
```suggestion
    @can_return_tuple
    @auto_docstring
```

File: src/transformers/models/radio/modular_radio.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    def forward(self, pixel_values: torch.Tensor, **kwargs) -> RadioModelOutput:
+        pixel_values = self.input_conditioner(pixel_values)
+        hidden_states = self.embeddings(pixel_values)
+        hidden_states = self.encoder(hidden_states, **kwargs)
+
+        num_skip = self.config.num_summary_tokens
+        all_summary = hidden_states[:, : self.config.num_cls_tokens]
+        summary = all_summary[:, self.summary_idxs].flatten(1)
+        features = hidden_states[:, num_skip:]
+
```
Comment (ArthurZucker) [inline src/transformers/models/radio/modular_radio.py]: is there no loss here we can add?

File: tests/models/radio/test_modeling_radio.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    def test_feed_forward_chunking(self):
+        pass
+
+    @unittest.skip(reason="RADIOModel uses pixel_values, not token embeddings")
+    def test_model_get_set_embeddings(self):
+        pass
+
+    @unittest.skip(reason="RadioModelOutput does not carry per-layer hidden_states")
+    def test_hidden_states_output(self):
+        pass
```
Comment (ArthurZucker) [inline tests/models/radio/test_modeling_radio.py]: this one just needs `OuptutRecorder`

File: tests/models/radio/test_modeling_radio.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+    @unittest.skip(reason="RadioModelOutput does not carry per-layer hidden_states")
+    def test_retain_grad_hidden_states_attentions(self):
+        pass
+
+    @unittest.skip(reason="RADIOModel.forward does not support return_dict=False")
```
Comment (ArthurZucker) [inline tests/models/radio/test_modeling_radio.py]: it should !

File: tests/models/radio/test_modeling_radio.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    def test_model_outputs_equivalence(self):
+        pass
+
+    @unittest.skip(reason="norm_mean/norm_std/summary_idxs buffers are initialized in __init__, not _init_weights")
+    def test_init_weights_can_init_buffers(self):
+        pass
```
Comment (ArthurZucker) [inline tests/models/radio/test_modeling_radio.py]: should overload init_weights

File: tests/models/radio/test_modeling_radio.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+        self.assertFalse(outputs.summary.isnan().any(), "summary contains NaN")
+        self.assertFalse(outputs.features.isnan().any(), "features contain NaN")
+
+
+# Parameterized SDPA inference comparison tests access outputs["hidden_states"],
+# which RadioModelOutput does not provide. Override each inherited test with a skip.
+_skip_sdpa = unittest.skip("RadioModelOutput does not provide hidden_states")(lambda self, *a, **kw: None)
+for _name in [n for n in dir(ModelTesterMixin) if n.startswith("test_") and "sdpa_inference" in n]:
+    setattr(RADIOModelTest, _name, _skip_sdpa)
+del _name, _skip_sdpa
```
Comment (ArthurZucker) [inline tests/models/radio/test_modeling_radio.py]: this is not the way we skip them!

Comment (ArthurZucker) [reply tests/models/radio/test_modeling_radio.py]: also HS are important you can easily record them with output recorders
